### ISSUE TITLE
homme versions of e3sm and scream do not match

### ISSUE THREAD

**Issue by oksanaguba** [E3SM-Project/scream#1816]
info on commits from e3sm and scream clones is below (just a sanity check), there is nothing important in git status for both clones either.

```
[ac.onguba@chrlogin2 scream-master]$ cd ~/scream-master/
[ac.onguba@chrlogin2 scream-master]$ git branch
  dyamond2
  jgfouca/homme_scream_flag_integration
  jgfouca/upstream_merge_2021_01_15
* master
  oksanaguba/eam/add-output
  oksanaguba/eam/add-pnh
  oksanaguba/eam/add-w-phi
[ac.onguba@chrlogin2 scream-master]$ git show
commit 55bebc98c9ed0bd37e2419fcfc65b4ba69443849 (HEAD -> master, origin/master, origin/HEAD)
Merge: d6e9e1d434 fc5f771126
Author: Autotester for E3SM related projects <56648600+E3SM-Autotester@users.noreply.github.com>
Date:   Thu Jul 21 18:21:16 2022 -0600

    Merge Pull Request #1809 from E3SM-Project/scream/aarondonahue/spa_bugfix_with_hya_hyb_coor
    
    Automatically Merged using E3SM Pull Request AutoTester
    PR Title: Remove redundant hyam and hybm variables from SPA interface.
    PR Author: AaronDonahue

[ac.onguba@chrlogin2 scream-master]$ 
[ac.onguba@chrlogin2 scream-master]$ cd ~/acme-master/
[ac.onguba@chrlogin2 acme-master]$ git branch
  azamat/chrysalis/add-prod-pelayouts
  azamat/perf/add-bench-tests
  maint-1.0
* master
  next
  oksanaguba/asteyer-fix
  oksanaguba/clubb-shflx
  oksanaguba/eam/add-output
  oksanaguba/eam/energy-diagn2
  oksanaguba/eam/energy-diagn3
  oksanaguba/eam/energy3
  oksanaguba/eam/fix-nh-restart-bug
  oksanaguba/eam/nltests
  oksanaguba/eam/tests-ftypes
  oksanaguba/homme/asteyer-fix
  oksanaguba/homme/bubble-rjmods
  oksanaguba/homme/fix-caar-test
  oksanaguba/homme/planar-uneven
  oksanaguba/homme/print-diagn
  oksanaguba/homme/vsplit2
  oksanaguba/homme/w-output
  temp
[ac.onguba@chrlogin2 acme-master]$ git show
commit 12bda05132daa2e7f43e2dfc6c1d628926560527 (HEAD -> master, origin/master, origin/HEAD)
Merge: 6d9ab1aa6e 98f272ac1e
Author: James Foucar <jgfouca@sandia.gov>
Date:   Thu Jul 21 11:38:33 2022 -0600

    Merge branch 'jgfouca/cime_update_2022_07_20' into master (PR #5088)
    
    ... to a00a6ed2fb2e8a8dceedc3973585d7cc3d867707
    
    Changes:
    1) Allow models to own their own config_files file
    2) Add custom PE file option to Jenkins jobs
    3) Add --pes-file option to bless_test_results
    4) Moving the statistical_ensemble_test files to CESM repo
    5) Add a new system test type, REUSEINITFILES
    6) Add new ParamGen utility
    
    Fixes:
    1) Changes source command from source to . for sh shell, leaves source for csh.
    2) Fix ww3dev baseline generation
    3) Fixes setting correct path when creating testreporter from the testreporter.template template
    
    [BFB]

```

**Comment by oksanaguba** [comment]
The diffs are:

```
[ac.onguba@chrlogin2 ~]$ diff -r ~/scream-master/components/homme/src/ ~/acme-master/components/homme/src/
Only in /home/ac.onguba/acme-master/components/homme/src/: aaa
Only in /home/ac.onguba/acme-master/components/homme/src/: aa.py
diff -r /home/ac.onguba/scream-master/components/homme/src/preqx_kokkos/CMakeLists.txt /home/ac.onguba/acme-master/components/homme/src/preqx_kokkos/CMakeLists.txt
22a23
>     ${TARGET_DIR}/cxx
26d26
<     ${TARGET_DIR}/cxx
28c28
<     ${HOMME_BINARY_DIR}/src/share/cxx)
---
>     ${HOMME_BINARY_DIR}/src/share/cxx
141,143d140
<     ${SRC_DIR}/repro_sum_mod.F90
<     ${SRC_DIR}/surfaces_mod.F90
<     ${SRC_DIR}/test_mod.F90
151a149
>     ${SRC_DIR}/repro_sum_mod.F90
152a151,152
>     ${SRC_DIR}/surfaces_mod.F90
>     ${SRC_DIR}/test_mod.F90
238c238
<   IF (PREQX_QSIZE_D)
---
>   IF (PREQX_QSIZE_D)      
diff -r /home/ac.onguba/scream-master/components/homme/src/preqx_kokkos/prim_driver_mod.F90 /home/ac.onguba/acme-master/components/homme/src/preqx_kokkos/prim_driver_mod.F90
45,80d44
<     !Init the kokkos functors (and their boundary exchanges)
<     call prim_init_kokkos_functors ()
< 
<     ! Init the kokkos views
<     call prim_init_elements_views (elem)
<   end subroutine prim_init2
< 
<   subroutine prim_create_c_data_structures (tl, hvcoord, mp)
<     use iso_c_binding,    only : c_loc, c_ptr, c_bool, C_NULL_CHAR
<     use time_mod,         only : TimeLevel_t
<     use hybvcoord_mod,    only : hvcoord_t
<     use prim_driver_base, only : deriv1
<     use control_mod,      only : vert_remap_q_alg, use_cpstar, transport_alg,           &
<                                  tstep_type, statefreq, rsplit, qsplit, ftype,          &
<                                  prescribed_wind, limiter_option, disable_diagnostics,  &
<                                  nu, nu_p, nu_q, nu_s, nu_div, nu_top, moisture,        &
<                                  hypervis_order, hypervis_scaling, hypervis_subcycle
<     use preqx_f2c_mod,    only : init_reference_element_c, init_simulation_params_c, &
<                                  init_hvcoord_c, init_time_level_c, init_elements_c
<     !
<     ! Input(s)
<     !
<     type (TimeLevel_t),       intent(in) :: tl
<     type (hvcoord_t), target, intent(in) :: hvcoord
<     real(kind=real_kind),     intent(in) :: mp(np,np)
<     !
<     ! Local(s)
<     !
<     logical (kind=c_bool) :: use_semi_lagrange_transport
<     real (kind=real_kind), dimension(np,np), target :: dvv, elem_mp
< 
<     type (c_ptr) :: hybrid_am_ptr, hybrid_ai_ptr, hybrid_bm_ptr, hybrid_bi_ptr
< 
<     ! Init the c data structures
<     call prim_create_c_data_structures(tl,hvcoord,elem(1)%mp)
< 
176,209d139
<   end subroutine prim_create_c_data_structures
< 
<   subroutine prim_init_elements_views (elem)
<     use iso_c_binding, only : c_ptr, c_loc
<     use element_mod,   only : element_t
<     use element_state, only : elem_state_v, elem_state_temp, elem_state_dp3d,       &
<                               elem_state_Q, elem_state_Qdp, elem_state_ps_v,        &
<                               elem_accum_qvar, elem_accum_qmass, elem_accum_q1mass, &
<                               elem_accum_kener, elem_accum_pener, elem_accum_iener, &
<                               elem_accum_iener_wet
<     use preqx_f2c_mod, only : init_elements_2d_c, init_elements_states_c, init_diagnostics_c
<     !
<     ! Input(s)
<     !
<     type (element_t), intent(in) :: elem(:)
<     !
<     ! Local(s)
<     !
<     type (c_ptr) :: elem_D_ptr, elem_Dinv_ptr
<     type (c_ptr) :: elem_spheremp_ptr, elem_rspheremp_ptr, elem_metdet_ptr
<     type (c_ptr) :: elem_metinv_ptr, elem_tensorvisc_ptr, elem_vec_sph2cart_ptr
<     type (c_ptr) :: elem_state_phis_ptr, elem_fcor_ptr
<     type (c_ptr) :: elem_state_ps_v_ptr, elem_state_dp3d_ptr
<     type (c_ptr) :: elem_state_v_ptr, elem_state_temp_ptr
<     type (c_ptr) :: elem_state_q_ptr, elem_state_Qdp_ptr
<     type (c_ptr) :: elem_accum_qvar_ptr, elem_accum_qmass_ptr, elem_accum_q1mass_ptr
<     type (c_ptr) :: elem_accum_iener_ptr, elem_accum_iener_wet_ptr, elem_accum_kener_ptr, elem_accum_pener_ptr
< 
<     real (kind=real_kind), target, dimension(np,np,2,2)     :: elem_D, elem_Dinv, elem_metinv, elem_tensorvisc
<     real (kind=real_kind), target, dimension(np,np)         :: elem_spheremp, elem_rspheremp, elem_metdet
<     real (kind=real_kind), target, dimension(np,np)         :: elem_state_phis, elem_fcor
<     real (kind=real_kind), target, dimension(np,np,3,2)     :: elem_vec_sph2cart
<     integer :: ie
< 
220a151
> 
diff -r /home/ac.onguba/scream-master/components/homme/src/share/cxx/ElementsGeometry.cpp /home/ac.onguba/acme-master/components/homme/src/share/cxx/ElementsGeometry.cpp
64,86d63
< set_phis (const int ie, CF90Ptr& phis) {
<   // Check geometry was inited
<   assert (m_num_elems>0);
< 
<   // Check input
<   assert (ie>=0 && ie<m_num_elems);
< 
<   using ScalarView    = ExecViewUnmanaged<Real [NP][NP]>;
<   using ScalarViewF90 = HostViewUnmanaged<const Real [NP][NP]>;
< 
<   ScalarViewF90           h_phis_f90 (phis);
<   ScalarView::HostMirror  h_phis = Kokkos::create_mirror_view(Homme::subview(m_phis,ie));
< 
<   for (int igp = 0; igp < NP; ++igp) {
<     for (int jgp = 0; jgp < NP; ++jgp) {
<       h_phis (igp, jgp) = h_phis_f90 (igp,jgp);
<     }
<   }
< 
<   Kokkos::deep_copy(Homme::subview(m_phis,ie), h_phis);
< }
< 
< void ElementsGeometry::
195a173,195
> }
> 
> void ElementsGeometry::
> set_phis (const int ie, CF90Ptr& phis) {
>   // Check geometry was inited
>   assert (m_num_elems>0);
> 
>   // Check input
>   assert (ie>=0 && ie<m_num_elems);
> 
>   using ScalarView    = ExecViewUnmanaged<Real [NP][NP]>;
>   using ScalarViewF90 = HostViewUnmanaged<const Real [NP][NP]>;
> 
>   ScalarViewF90           h_phis_f90 (phis);
>   ScalarView::HostMirror  h_phis = Kokkos::create_mirror_view(Homme::subview(m_phis,ie));
> 
>   for (int igp = 0; igp < NP; ++igp) {
>     for (int jgp = 0; jgp < NP; ++jgp) {
>       h_phis (igp, jgp) = h_phis_f90 (igp,jgp);
>     }
>   }
> 
>   Kokkos::deep_copy(Homme::subview(m_phis,ie), h_phis);
diff -r /home/ac.onguba/scream-master/components/homme/src/share/cxx/ElementsGeometry.hpp /home/ac.onguba/acme-master/components/homme/src/share/cxx/ElementsGeometry.hpp
71a72
> 
diff -r /home/ac.onguba/scream-master/components/homme/src/share/cxx/FunctorsBuffersManager.cpp /home/ac.onguba/acme-master/components/homme/src/share/cxx/FunctorsBuffersManager.cpp
58d57
<   generate_random_data();
diff -r /home/ac.onguba/scream-master/components/homme/src/share/cxx/Hommexx_Session.cpp /home/ac.onguba/acme-master/components/homme/src/share/cxx/Hommexx_Session.cpp
46,49c46,49
<   std::cout << "HOMMEXX VECTOR_SIZE: " << VECTOR_SIZE << "\n";
<   std::cout << "HOMMEXX vector tag: " << Scalar::label() << "\n";
<   std::cout << "HOMMEXX active AVX set:" << active_avx_string() << "\n";
<   std::cout << "HOMMEXX MPI_ON_DEVICE: " << HOMMEXX_MPI_ON_DEVICE << "\n";
---
>     std::cout << "HOMMEXX VECTOR_SIZE: " << VECTOR_SIZE << "\n";
>     std::cout << "HOMMEXX vector tag: " << Scalar::label() << "\n";
>     std::cout << "HOMMEXX active AVX set:" << active_avx_string() << "\n";
>     std::cout << "HOMMEXX MPI_ON_DEVICE: " << HOMMEXX_MPI_ON_DEVICE << "\n";
51c51
<   std::cout << "HOMMEXX CUDA_SHARE_BUFFER: on\n";
---
>     std::cout << "HOMMEXX CUDA_SHARE_BUFFER: on\n";
53c53
<   std::cout << "HOMMEXX CUDA_SHARE_BUFFER: off\n";
---
>     std::cout << "HOMMEXX CUDA_SHARE_BUFFER: off\n";
55,56c55,56
<   std::cout << "HOMMEXX CUDA_(MIN/MAX)_WARP_PER_TEAM: " << HOMMEXX_CUDA_MIN_WARP_PER_TEAM
<             << " / " << HOMMEXX_CUDA_MAX_WARP_PER_TEAM << "\n";
---
>     std::cout << "HOMMEXX CUDA_(MIN/MAX)_WARP_PER_TEAM: " << HOMMEXX_CUDA_MIN_WARP_PER_TEAM
>               << " / " << HOMMEXX_CUDA_MAX_WARP_PER_TEAM << "\n";
diff -r /home/ac.onguba/scream-master/components/homme/src/share/gllfvremap_mod.F90 /home/ac.onguba/acme-master/components/homme/src/share/gllfvremap_mod.F90
2561c2561
<     integer :: nets, nete, nerr
---
>     integer :: nets, nete, nerr, ic
2730a2731
>        ic = 2
2735c2736
<                sum(wg*(elem(ie)%state%Q(:,:,1,1) - elem(ie)%state%Q(:,:,1,2))**2)
---
>                sum(wg*(elem(ie)%state%Q(:,:,1,1) - elem(ie)%state%Q(:,:,1,ic))**2)
2737c2738
<                sum(wg*elem(ie)%state%Q(:,:,1,2)**2)
---
>                sum(wg*elem(ie)%state%Q(:,:,1,ic)**2)
2740c2741
<           global_shared_buf(ie,3) = sum(wg*elem(ie)%state%Q(:,:,1,2))
---
>           global_shared_buf(ie,3) = sum(wg*elem(ie)%state%Q(:,:,1,ic))
2743c2744
<           qmin1 = min(qmin1, minval(elem(ie)%state%Q(:,:,1,2)))
---
>           qmin1 = min(qmin1, minval(elem(ie)%state%Q(:,:,1,ic)))
2745c2746
<           qmax1 = max(qmax1, maxval(elem(ie)%state%Q(:,:,1,2)))
---
>           qmax1 = max(qmax1, maxval(elem(ie)%state%Q(:,:,1,ic)))
diff -r /home/ac.onguba/scream-master/components/homme/src/share/namelist_mod.F90 /home/ac.onguba/acme-master/components/homme/src/share/namelist_mod.F90
221c221
<     character(len=MAX_STRING_LEN) :: se_topology, se_geometry
---
>     character(len=MAX_STRING_LEN) :: se_topology
224,225d223
<     integer :: se_ne_x, se_ne_y
<     real(kind=real_kind) :: se_lx, se_ly
241d238
<       se_geometry,       &
243,246d239
<       se_ne_x,           &
<       se_ne_y,           &
<       se_lx,             &
<       se_ly,             &
706,709d698
<        ne_x       = se_ne_x
<        ne_y       = se_ne_y
<        Lx         = se_lx
<        Ly         = se_ly
711d699
<        geometry   = se_geometry
747,748d734
<     call MPI_bcast(Lx,              1, MPIreal_t,par%root,par%comm,ierr)
<     call MPI_bcast(Ly,              1, MPIreal_t,par%root,par%comm,ierr)
965c951
< if (topology == "plane" .and. (mesh_file /= "/dev/null" .and. mesh_file /= "none")) then
---
> if (topology == "plane" .and. mesh_file /= "none") then
1006,1008d991
<       Sx = 0.0D0
<       Sy = 0.0D0
< 
1023,1028c1006
<       if (Lx==0.0 .or. Ly==0.0) then
<          print *, 'For planar homme Lx and Ly cannot be zero'
<          call abortmp("Error Lx or Ly = 0")
<       endif
< 
<    else if (geometry == "sphere") then
---
>     else if (geometry == "sphere") then
diff -r /home/ac.onguba/scream-master/components/homme/src/theta-l/share/element_ops.F90 /home/ac.onguba/acme-master/components/homme/src/theta-l/share/element_ops.F90
70c70
<   public get_field, get_field_i, get_state, get_pottemp
---
>   public get_field, get_field_i, get_state
diff -r /home/ac.onguba/scream-master/components/homme/src/theta-l_kokkos/CMakeLists.txt /home/ac.onguba/acme-master/components/homme/src/theta-l_kokkos/CMakeLists.txt
1a2
> 
14,15c15,16
<   SET(UTILS_TIMING_SRC_DIR  ${HOMME_SOURCE_DIR}/utils/cime/CIME/non_py/src/timing)
<   SET(UTILS_TIMING_DIR      ${HOMME_BINARY_DIR}/utils/cime/CIME/non_py/src/timing)
---
>   SET(UTILS_TIMING_SRC_DIR  ${CMAKE_SOURCE_DIR}/utils/cime/CIME/non_py/src/timing)
>   SET(UTILS_TIMING_DIR      ${CMAKE_BINARY_DIR}/utils/cime/CIME/non_py/src/timing)
20d20
<     ${UTILS_TIMING_DIR}
21a22
>     ${UTILS_TIMING_DIR}
24c25
<     ${HOMME_BINARY_DIR}/src/share/cxx
---
>     ${CMAKE_BINARY_DIR}/src/share/cxx
132c133
<     ${SRC_DIR}/checksum_mod.F90
---
>     ${SRC_DIR}/checksum_mod.F90 
146,150c147,151
<     ${UTILS_SHARE_DIR}/shr_mpi_mod.F90
<     ${UTILS_SHARE_DIR}/shr_sys_mod.F90
<     ${UTILS_SHARE_DIR}/shr_file_mod.F90
<     ${UTILS_SHARE_DIR}/shr_vmath_mod.F90
<     ${UTILS_SHARE_DIR}/shr_const_mod.F90
---
>     ${UTILS_SHARE_DIR}/shr_mpi_mod.F90 
>     ${UTILS_SHARE_DIR}/shr_sys_mod.F90 
>     ${UTILS_SHARE_DIR}/shr_file_mod.F90 
>     ${UTILS_SHARE_DIR}/shr_vmath_mod.F90 
>     ${UTILS_SHARE_DIR}/shr_const_mod.F90 
diff -r /home/ac.onguba/scream-master/components/homme/src/theta-l_kokkos/cxx/CaarFunctorImpl.hpp /home/ac.onguba/acme-master/components/homme/src/theta-l_kokkos/cxx/CaarFunctorImpl.hpp
115d114
<   Kokkos::RangePolicy<ExecSpace, TagPostExchange> m_policy_post;
116a116,117
> 
>   Kokkos::RangePolicy<ExecSpace, TagPostExchange> m_policy_post;
diff -r /home/ac.onguba/scream-master/components/homme/src/theta-l_kokkos/prim_driver_mod.F90 /home/ac.onguba/acme-master/components/homme/src/theta-l_kokkos/prim_driver_mod.F90
5a6,7
>   use prim_driver_base,     only: deriv1, smooth_topo_datasets
>   use prim_cxx_driver_base, only: prim_init1, prim_finalize
470a473
> 
diff -r /home/ac.onguba/scream-master/components/homme/src/theta-l_kokkos/share/element_ops.F90 /home/ac.onguba/acme-master/components/homme/src/theta-l_kokkos/share/element_ops.F90
70c70
<   public get_field, get_field_i, get_state, get_pottemp
---
>   public get_field, get_field_i, get_state
[ac.onguba@chrlogin2 ~]$ 
```

**Comment by ambrad** [comment]
The diffs in gllfvremap_mod.F90 will be resolved after the next upstream merge.

**Comment by bartgol** [comment]
I think the `namelist_mod.F90` diffs are due to doubly periodic scream stuff. Not sure if it makes sense to put it in the upstream before scream is merged back into e3sm. The diff in `share/cxx/FunctorsBuffersManager.cpp` is worrysome. The extra line in scream's repo should cause an infinite recursion! We never use it in scream, but we better remove that line before we merge into the upstream. Some diffs are extra empty lines and/or function definitions rearranged and/or different indentation. Can be fixed in scream, to match upstream. `prim_driver_mod.F90` in preqx seems utterly wrong in the scream repo. Probably a merge gone wrong.

Since I probably created the mess, I can clean up everything (except `namelist_mod.F90`) a mini branch on the scream side. But I can wait until Andrew is done, in case there would be conflicts.

**Comment by ambrad** [comment]
Thanks @bartgol, but my upstream changes don't touch these files, so you can reconcile the diffs whenever you want. (Just for clarity and repeating my comment above, gllfvremap_mod is not a true discrepancy, so please don't change that file; SCREAM just hasn't gotten the latest commit to that file from upstream yet.)

**Comment by oksanaguba** [comment]
To add, there is a typo in preqx_kokkos cmake:
https://github.com/E3SM-Project/scream/blob/5677efbeea579c71ed54fdb2bfaf72ece7cad51c/components/homme/src/preqx_kokkos/CMakeLists.txt#L27

**Comment by oksanaguba** [comment]
This was resolved partially, i still need to resolve conflicts around DP namelist vars.

**Comment by ambrad** [comment]
I'll take over this task because I'm going to work on the C++ DP capability in Hommexx once the RRM task is done.

**Comment by ambrad** [comment]
The namelist_mod part will be resolved with the SCREAM->E3SM merge PR.
